# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/ipc/RDDChild.cpp
# Commit: 493b443954fe
# Full Hash: 493b443954fe15f7b542ba14671f25e5f8531dff
# Author: Jed Davis <jld@mozilla.com>
# Date: 2019-02-28 10:55:23
# Regressor Bug: 1506291
# File Overlap Count: 1
# Description:
#   Bug 1506291 - Add Linux sandboxing for the RDD (media decoder) process. r=gcp,mjf,flod
#   
#   The seccomp-bpf policy is currently just the "common" policy with no
#   additions (but with the fixes in bug 1511560 to enable shared memory
#   creation).  The file broker policy allows shared memory creation and
# ==============================================================================

diff -r bf58d8320f5a -r 493b443954fe dom/media/ipc/RDDChild.cpp
--- a/dom/media/ipc/RDDChild.cpp	Mon Feb 25 16:20:50 2019 +0000
+++ b/dom/media/ipc/RDDChild.cpp	Wed Feb 27 20:14:54 2019 +0000
@@ -8,6 +8,11 @@
 #include "mozilla/dom/MemoryReportRequest.h"
 #include "mozilla/ipc/CrashReporterHost.h"
 
+#if defined(XP_LINUX) && defined(MOZ_SANDBOX)
+#  include "mozilla/SandboxBroker.h"
+#  include "mozilla/SandboxBrokerPolicyFactory.h"
+#endif
+
 #ifdef MOZ_GECKO_PROFILER
 #  include "ProfilerParent.h"
 #endif
@@ -23,12 +28,31 @@
 
 RDDChild::~RDDChild() { MOZ_COUNT_DTOR(RDDChild); }
 
-void RDDChild::Init() {
-  SendInit();
+bool RDDChild::Init() {
+  MaybeFileDesc brokerFd = void_t();
+
+#if defined(XP_LINUX) && defined(MOZ_SANDBOX)
+  auto policy = SandboxBrokerPolicyFactory::GetUtilityPolicy(OtherPid());
+  if (policy != nullptr) {
+    brokerFd = FileDescriptor();
+    mSandboxBroker =
+        SandboxBroker::Create(std::move(policy), OtherPid(), brokerFd);
+    // This is unlikely to fail and probably indicates OS resource
+    // exhaustion, but we can at least try to recover.
+    if (NS_WARN_IF(mSandboxBroker == nullptr)) {
+      return false;
+    }
+    MOZ_ASSERT(static_cast<const FileDescriptor&>(brokerFd).IsValid());
+  }
+#endif  // XP_LINUX && MOZ_SANDBOX
+
+  SendInit(brokerFd);
 
 #ifdef MOZ_GECKO_PROFILER
   Unused << SendInitProfiler(ProfilerParent::CreateForProcess(OtherPid()));
 #endif
+
+  return true;
 }
 
 bool RDDChild::EnsureRDDReady() {